assertj-core/src/main/java/org/assertj/core/api/AbstractIteratorAssert.java

```diff
@@ -93,4 +96,49 @@ public IterableAssert<ELEMENT> toIterable() {
     return new IterableAssert<>(IterableAssert.toIterable(actual));
   }
 
+  /**
+   * Verifies that the actual iterator is unmodifiable, i.e., throws an {@link UnsupportedOperationException} with
+   * any attempt to remove from the iterator.
+   * <p>
+   * Example:
+   * <pre><code class='java'> // assertions will pass
+   * assertThat(List.of().iterator()).isUnmodifiable();
+   * assertThat(Set.of().iterator()).isUnmodifiable();
+   *
+   * // assertions will fail
+   * assertThat(new ArrayList&lt;&gt;().iterator()).isUnmodifiable();
+   * assertThat(new HashSet&lt;&gt;().iterator()).isUnmodifiable();</code></pre>
+   *
+   * @return {@code this} assertion object.
+   * @throws AssertionError if the actual iterator is modifiable.
+   */
+  @Beta
+  public SELF isUnmodifiable() {
+    isNotNull();
+    assertIsUnmodifiable();
+    return myself;
+  }
+
+  private void assertIsUnmodifiable() {
+    switch (actual.getClass().getName()) {
+      case "java.util.Collections$EmptyIterator":
+      case "java.util.Collections$EmptyListIterator":
+        // immutable by contract, although not all methods throw UnsupportedOperationException
+        return;
+    }
+
+    expectUnsupportedOperationException(() -> actual.remove(), "Iterator.remove()");
```

Comment:
> But can't rely on `IllegalStateException` as unmodifiability marker. `ArrayList<>().itertator()` can also throw `IllegalStateException` if we haven't called `next`.
> 
> Initially, I wanted to check unmodifiability by just calling `iterator.remove()`. It's not enough (see to ArrayList example). Now, I'm thinking of making an additional check like `iterator.next(); iterator.remove()`. However, it will look a bit weird.

According to the [`remove()`](https://docs.oracle.com/javase/8/docs/api/java/util/Iterator.html#remove--) contract, a preliminary call to `next()` is required (probably to be guarded by `hasNext()`). We might go this way and see how the modifiable implementations behave. WDYT?

Reply:
With the same thoughts, I wrote my comment in the ticket :) 
I thought of making two checks and expecting either `remove()` or `next(); remove()` will throw `UnsupportedOperationException`. I couldn't implement it nicely, which made me think that I'm overenginerring. 
Then I thought that if an iterator is unmodifiable, whyever on the earth somebody will implement any other logic in `remove()` except just throwing an exception? o_O That's why I left only this check.

However if you think it makes sense to call `next()` before `remove()` I can add it, ofc with guarded
`hasNext()`.  So I have three options for you:
1. Accept my argument to common sense and leave it as it is;
2. Call `next(); remove()` in the current check;
3. Add two checks, one with just `remove()` and another with `next(); remove()`;
4. Ofc I am open to any other suggestions ;) 

so, WDYT?

Reply:
Let's be pragmatic and go with option 1!